File: src/main/java/tech/bobliu/assignment07/controller/UserSessionListener.java
package tech.bobliu.assignment07.controller;

import jakarta.servlet.annotation.WebListener;
import jakarta.servlet.http.HttpSessionEvent;
import jakarta.servlet.http.HttpSessionListener;
import tech.bobliu.assignment07.model.User;
import tech.bobliu.assignment07.service.UserService;

@WebListener
public class UserSessionListener implements HttpSessionListener {
    @Override
    public void sessionDestroyed(HttpSessionEvent se) {
        Object obj = se.getSession().getAttribute("user");
        if (obj instanceof User user) {
            UserService.setUserAsOffline(user);
        }
    }
}

File: src/main/kotlin/tech/bobliu/assignment07/controller/UserSessionListener.kt
package tech.bobliu.assignment07.controller

import jakarta.servlet.annotation.WebListener
import jakarta.servlet.http.HttpSessionEvent
import jakarta.servlet.http.HttpSessionListener
import tech.bobliu.assignment07.model.User
import tech.bobliu.assignment07.service.UserService

@WebListener
class UserSessionListener : HttpSessionListener {
    override fun sessionDestroyed(se: HttpSessionEvent) {
        (se.session.getAttribute("user") as? User)?.let { user ->
            UserService.setUserAsOffline(user)
        }
    }
}